sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala

```diff
@@ -300,97 +301,108 @@ final class DataStreamWriter[T] private[sql](ds: Dataset[T]) {
+        }
+        val sink = new MemorySink()
+        val resultDf = Dataset.ofRows(df.sparkSession, new MemoryPlan(sink, df.schema.toAttributes))
+
+        (sink, Some(resultDf), outputMode == OutputMode.Complete(), extraOptions)
+      } else if (source == SOURCE_NAME_FOREACH) {
+        assertNotPartitioned("foreach")
+        val sink = ForeachWriterTable[T](foreachWriter, ds.exprEnc)
+        (sink, None, true, extraOptions)
+      } else if (source == SOURCE_NAME_FOREACH_BATCH) {
+        assertNotPartitioned("foreachBatch")
+        if (trigger.isInstanceOf[ContinuousTrigger]) {
+          throw new AnalysisException("'foreachBatch' is not supported with continuous trigger")
         }
+        val sink = new ForeachBatchSink[T](foreachBatchWriter, ds.exprEnc)
+        (sink, None, true, extraOptions)
       } else {
-        createV1Sink(optionsWithPath)
-      }
+        val cls = DataSource.lookupDataSource(source, df.sparkSession.sessionState.conf)
+        val disabledSources = df.sparkSession.sqlContext.conf.disabledV2StreamingWriters.split(",")
+        val useV1Source = disabledSources.contains(cls.getCanonicalName) ||
+          // file source v2 does not support streaming yet.
+          classOf[FileDataSourceV2].isAssignableFrom(cls)
+
+        val optionsWithPath = if (path.isEmpty) {
+          extraOptions
+        } else {
+          extraOptions + ("path" -> path.get)
+        }
+
+        val sink = if (classOf[TableProvider].isAssignableFrom(cls) && !useV1Source) {
+          val provider = cls.getConstructor().newInstance().asInstanceOf[TableProvider]
+          val sessionOptions = DataSourceV2Utils.extractSessionConfigs(
+            source = provider, conf = df.sparkSession.sessionState.conf)
+          val finalOptions = sessionOptions.filterKeys(!optionsWithPath.contains(_)).toMap ++
+            optionsWithPath.originalMap
+          val dsOptions = new CaseInsensitiveStringMap(finalOptions.asJava)
+          val table = DataSourceV2Utils.getTableFromProvider(
+            provider, dsOptions, userSpecifiedSchema = None)
+          import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Implicits._
+          table match {
+            case table: SupportsWrite if table.supports(STREAMING_WRITE) =>
+              table
+            case _ => createV1Sink(optionsWithPath)
+          }
+        } else {
+          createV1Sink(optionsWithPath)
+        }
 
-      df.sparkSession.sessionState.streamingQueryManager.startQuery(
-        extraOptions.get("queryName"),
-        extraOptions.get("checkpointLocation"),
-        df,
-        optionsWithPath.originalMap,
-        sink,
-        outputMode,
-        useTempCheckpointLocation = source == "console" || source == "noop",
-        recoverFromCheckpointLocation = true,
-        trigger = trigger)
+        (sink, None, true, optionsWithPath.originalMap)
+      }
     }
+
+    val query = df.sparkSession.sessionState.streamingQueryManager.startQuery(
```

Comment:
another way to refactor:
```
private def startQuery(sink, resultDf, recoverFromCheckpoint, newOptions) = {
  df.sparkSession.sessionState.streamingQueryManager.startQuery...
}

def start() = {
  if (source == SOURCE_NAME_TABLE) = {
    ...
    startQuery...
  } else if (...) {
    ...
    startQuery...
  } ...
}
```

this should reduce code diff quite a bit.

Reply:
I'm OK on the refactored result in either way. I'll try it and see the code diff. Thanks!